File: app/src/main/java/com/accentury/app/ui/components/AccenturyButton.kt
package com.accentury.app.ui.components

import androidx.compose.animation.core.animateFloatAsState
import androidx.compose.animation.core.tween
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.interaction.MutableInteractionSource
import androidx.compose.foundation.interaction.collectIsPressedAsState
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.defaultMinSize
import androidx.compose.foundation.layout.offset
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.alpha
import androidx.compose.ui.draw.clip
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.compose.ui.unit.times
import com.accentury.app.ui.theme.Dimens
import com.accentury.app.ui.theme.Motion
import com.accentury.app.ui.theme.Radius
import com.accentury.app.ui.theme.Spacing
import com.accentury.app.ui.theme.accenturyColors
import com.accentury.app.ui.theme.motionDuration

/**
 * 버튼 무게 (KAN-148). 화면들이 실제로 쓰는 셋만 둔다.
 *
 * - [Primary] 주동작 (녹음 시작·다음·허용). 화면에서 유일하게 잉크로 꽉 찬 면이다
 * - [Secondary] 보조 (재녹음·재시도). 테두리만 두르고 그림자를 두지 않는다
 * - [Text] 이탈·종료. 눌리면 안 되는 쪽이라 무게를 가장 뺀다
 */
enum class ButtonVariant { Primary, Secondary, Text }

/**
 * 공통 버튼 (KAN-148, 형태는 KAN-161 2단계). 웹의 `.btn`과 같은 모양·같은 값이다 —
 * 두 런타임이 한 테스트 안에서 번갈아 나오므로 버튼이 서로 다르게 생기면 바로 보인다.
 *
 * 오려 낸 종이다: 오른쪽·아래로 어긋난 자리에 [paperShadow]가 단색 면 한 겹을 깔아 종이가
 * 떠 있는 것처럼 보이고, 누르면 본체가 정확히 그만큼 내려가 그림자를 덮는다 — 종이가
 * 바닥에 닿는 순간이다. 총 차지 높이는 눌림 전후로 같아서 옆 요소가 밀리지 않는다.
 *
 * 그림자는 주 버튼에만 있다. 그림자는 "떠 있다"는 뜻이라 화면에 떠 있는 종이가 둘이면
 * 어느 쪽을 눌러야 하는지가 흐려진다. 보조 버튼은 자리만 같게 비우고(눌림 거리가 같다)
 * 그림자를 그리지 않는다.
 *
 * 최소 높이는 주 버튼이 [Dimens.controlHeightLg] 56dp, 보조가 [Dimens.touchTargetMin] 48dp다
 * (ux-ui.md §5의 48dp 최소선을 둘 다 넘는다).
 */
@Composable
fun AccenturyButton(
    text: String,
    onClick: () -> Unit,
    modifier: Modifier = Modifier,
    variant: ButtonVariant = ButtonVariant.Primary,
    enabled: Boolean = true,
) {
    if (variant == ButtonVariant.Text) {
        TextButton(
            onClick = onClick,
            enabled = enabled,
            modifier = modifier.defaultMinSize(minHeight = Dimens.touchTargetMin),
        ) {
            Text(text, style = MaterialTheme.typography.labelLarge)
        }
        return
    }

    val colors = MaterialTheme.accenturyColors
    val interaction = remember { MutableInteractionSource() }
    val pressed by interaction.collectIsPressedAsState()

    val isPrimary = variant == ButtonVariant.Primary
    val shape = RoundedCornerShape(Radius.md)

    // 눌림은 0..1 한 값이다 - 본체가 내려가는 거리를 x·y 따로 애니메이션하면 두 축이
    // 미세하게 어긋나 종이가 비스듬히 미끄러진다.
    val sink by animateFloatAsState(
        targetValue = if (pressed && enabled) 1f else 0f,
        animationSpec = tween(
            durationMillis = motionDuration(Motion.PRESS),
            easing = Motion.easeOut,
        ),
        label = "buttonSink",
    )

    Box(
        modifier = modifier
            // 비활성은 불투명도만 낮춘다 - 회색으로 칠하면 색이 하나 더 늘고 배경에 따라
            // 대비를 잃는다. 원래 색을 흐리게 하면 대비가 함께 줄어 예측 가능하다.
            .alpha(if (enabled) 1f else DISABLED_ALPHA)
            .paperShadow(colors.primaryDim, Radius.md, visible = isPrimary)
            .offset(x = sink * Dimens.paperShadowX, y = sink * Dimens.paperShadowY)
            .defaultMinSize(
                minWidth = MIN_BUTTON_WIDTH,
                minHeight = if (isPrimary) Dimens.controlHeightLg else Dimens.touchTargetMin,
            )
            .clip(shape)
            .background(if (isPrimary) MaterialTheme.colorScheme.primary else MaterialTheme.colorScheme.surface)
            // 주 버튼도 테두리를 두른다. 면과 같은 색이라 낭비 같지만, 크림 배경 위에서
            // 잉크 면의 가장자리가 종이를 오린 자리처럼 또렷해진다.
            .border(
                width = if (isPrimary) PRIMARY_BORDER else SECONDARY_BORDER,
                color = colors.controlBorder,
                shape = shape,
            )
            .clickableButton(enabled, interaction, onClick)
            .padding(horizontal = Spacing.x6),
        contentAlignment = Alignment.Center,
    ) {
        if (isPrimary) {
            // 주 CTA 라벨은 Jua 20sp(`titleSmall`) - 이 화면에서 눌러야 할 것이 제목만큼 크다
            Text(
                text,
                style = MaterialTheme.typography.titleMedium,
                color = MaterialTheme.colorScheme.onPrimary,
                letterSpacing = PRIMARY_LETTER_SPACING,
            )
        } else {
            Text(
                text,
                style = MaterialTheme.typography.bodyMedium,
                fontWeight = FontWeight.Medium,
                color = MaterialTheme.colorScheme.onSecondaryContainer,
            )
        }
    }
}

/** 라벨이 짧아도 손가락이 닿을 만큼은 넓게 잡는다 */
private val MIN_BUTTON_WIDTH = 120.dp

/** 테두리 굵기. 주 CTA와 선택 상태만 2dp, 나머지는 1.5dp다 (시안 규칙) */
private val PRIMARY_BORDER = 2.dp
private val SECONDARY_BORDER = 1.5.dp

/** Jua 라벨의 자간. 굵기를 못 올리는 폰트라 자간이 라벨의 무게를 대신한다 */
private val PRIMARY_LETTER_SPACING = 0.4.sp

private const val DISABLED_ALPHA = 0.6f

/** ripple 없이 누름 상태만 받는다 - 눌림 표현은 종이가 내려가는 것으로 이미 하고 있다 */
private fun Modifier.clickableButton(
    enabled: Boolean,
    interaction: MutableInteractionSource,
    onClick: () -> Unit,
): Modifier = this.clickable(
    interactionSource = interaction,
    indication = null,
    enabled = enabled,
    onClick = onClick,
)

File: ios/Accentury/UI/Components/AccenturyButton.swift
import SwiftUI

/// 버튼 무게 (KAN-148). 안드로이드 `ButtonVariant`와 같은 셋이다.
///
/// - ``primary`` 주동작 (녹음 시작·다음·허용). 화면에서 유일하게 잉크로 꽉 찬 면이다
/// - ``secondary`` 보조 (재녹음·재시도). 테두리만 두르고 그림자를 두지 않는다
/// - ``text`` 이탈·복귀. 눌리면 안 되는 쪽이라 무게를 가장 뺀다
enum ButtonVariant {
    case primary
    case secondary
    case text
}

/// 공통 버튼 (KAN-148, 형태는 KAN-161 2단계). 안드로이드 `ui/components/AccenturyButton.kt`의
/// 이식본이고 웹의 `.btn`과 같은 모양·같은 값이다 — 두 런타임이 한 테스트 안에서 번갈아
/// 나오므로 버튼이 서로 다르게 생기면 바로 보인다.
///
/// 오려 낸 종이다: 오른쪽·아래로 어긋난 자리에 ``View/paperShadow(cornerRadius:visible:)``가
/// 단색 면 한 겹을 깔아 종이가 떠 있는 것처럼 보이고, 누르면 본체가 정확히 그만큼 내려가
/// 그림자를 덮는다 — 종이가 바닥에 닿는 순간이다. 총 차지 높이는 눌림 전후로 같아서 옆
/// 요소가 밀리지 않는다.
///
/// 그림자는 주 버튼에만 있다. 그림자는 "떠 있다"는 뜻이라 화면에 떠 있는 종이가 둘이면 어느
/// 쪽을 눌러야 하는지가 흐려진다. 보조 버튼은 자리만 같게 비우고(눌림 거리가 같다) 그림자를
/// 그리지 않는다.
///
/// 눌림을 제스처가 아니라 ``PapercutButtonStyle``로 받는다 — `Button`의 의미론(스크린 리더의
/// "버튼", 활성화)을 그대로 두면서 `isPressed`만 빌려 쓰는 자리가 그것이다. 제스처로 직접
/// 받으면 VoiceOver가 이 요소를 눌러 줄 방법이 사라진다.
struct AccenturyButton: View {

    let text: String
    var variant: ButtonVariant = .primary
    var enabled: Bool = true
    /// 폭을 늘려 잡을지. 안드로이드가 호출부에서 `Modifier.weight(1f)`/`fillMaxWidth()`로
    /// 정하던 자리다 — 검토 화면의 [재녹음]·[다음]이 같은 폭을 갖는 규칙이 여기 걸린다.
    var fillsWidth: Bool = false
    let action: () -> Void

    var body: some View {
        switch variant {
        case .text:
            Button(action: action) {
                Text(text)
                    .papercutType(.label)
                    .foregroundColor(Papercut.ink)
                    .frame(maxWidth: fillsWidth ? .infinity : nil)
                    .frame(minHeight: Papercut.touchTargetMin)
                    .padding(.horizontal, Papercut.space4)
                    .contentShape(Rectangle())
            }
            .buttonStyle(.plain)
            .disabled(!enabled)
            .opacity(enabled ? 1 : Papercut.opacityDisabled)

        case .primary, .secondary:
            Button(action: action) {
                // 주 CTA 라벨은 Jua 20(`title`)에 자간 0.4다 — 이 화면에서 눌러야 할 것이
                // 제목만큼 크고, 굵기를 못 올리는 폰트라 자간이 무게를 대신한다 (KAN-178).
                // 보조는 본문 글꼴 15(`bodySmall`)로, 안드로이드 `bodyMedium` 자리다.
                Text(text)
                    .papercutType(
                        variant == .primary ? .title : .bodySmall,
                        tracking: variant == .primary ? Papercut.primaryLabelTracking : nil
                    )
                    .foregroundColor(variant == .primary ? Papercut.cream : Papercut.ink)
            }
            .buttonStyle(PapercutButtonStyle(isPrimary: variant == .primary, fillsWidth: fillsWidth))
            .disabled(!enabled)
            // 비활성은 불투명도만 낮춘다 — 회색으로 칠하면 색이 하나 더 늘고 배경에 따라 대비를
            // 잃는다. 원래 색을 흐리게 하면 대비가 함께 줄어 예측 가능하다.
            .opacity(enabled ? 1 : Papercut.opacityDisabled)
        }
    }
}

/// 종이 버튼의 면·테두리·눌림. 라벨은 호출자가 그리고 여기는 그 라벨을 감싸는 종이만 그린다.
struct PapercutButtonStyle: ButtonStyle {

    let isPrimary: Bool
    var fillsWidth: Bool = false

    func makeBody(configuration: Configuration) -> some View {
        let shape = RoundedRectangle(cornerRadius: Papercut.radiusMD, style: .continuous)
        // 눌림은 0..1 한 값이다 — 본체가 내려가는 거리를 x·y 따로 애니메이션하면 두 축이
        // 미세하게 어긋나 종이가 비스듬히 미끄러진다.
        let sink: CGFloat = configuration.isPressed ? 1 : 0

        return configuration.label
            .padding(.horizontal, Papercut.space6)
            .frame(maxWidth: fillsWidth ? .infinity : nil)
            // 라벨이 짧아도 손가락이 닿을 만큼은 넓게 잡는다. 주 버튼이 56, 보조가 48이다 —
            // 둘의 무게 차이가 크기로도 읽힌다.
            .frame(minWidth: 120, minHeight: isPrimary ? Papercut.controlHeightLarge : Papercut.touchTargetMin)
            .background(shape.fill(isPrimary ? Papercut.ink : Papercut.cream))
            // 주 버튼도 테두리를 두른다. 면과 같은 색이라 낭비 같지만, 크림 배경 위에서 잉크
            // 면의 가장자리가 종이를 오린 자리처럼 또렷해진다.
            .overlay(shape.stroke(Papercut.ink, lineWidth: isPrimary ? Papercut.borderStrong : Papercut.borderRegular))
            .contentShape(shape)
            .offset(x: sink * Papercut.paperShadowX, y: sink * Papercut.paperShadowY)
            .paperShadow(cornerRadius: Papercut.radiusMD, visible: isPrimary)
            .animation(.easeOut(duration: Papercut.Motion.press), value: configuration.isPressed)
    }
}
